File: backend/src/main/java/com/hostel/backend/service/NoticeService.java
package com.hostel.backend.service;

import com.hostel.backend.dto.NoticeRequest;
import com.hostel.backend.dto.NoticeResponse;
import com.hostel.backend.entity.Notice;
import com.hostel.backend.entity.User;
import com.hostel.backend.enums.Role;
import com.hostel.backend.exception.AppException;
import com.hostel.backend.exception.ResourceNotFoundException;
import com.hostel.backend.exception.UnauthorizedException;
import com.hostel.backend.repository.NoticeRepository;
import com.hostel.backend.repository.UserRepository;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class NoticeService {

    private final NoticeRepository noticeRepository;
    private final UserRepository   userRepository;

    public NoticeService(NoticeRepository noticeRepository,
                         UserRepository userRepository) {
        this.noticeRepository = noticeRepository;
        this.userRepository   = userRepository;
    }

    @Transactional
    public NoticeResponse createNotice(String email, NoticeRequest request) {
        User user = getUser(email);

        if (user.getRole() != Role.ADMIN &&
            user.getRole() != Role.WARDEN &&
            user.getRole() != Role.CARETAKER) {
            throw new UnauthorizedException("Only admin, warden or caretaker can post notices");
        }
        if (request.getTitle() == null || request.getTitle().trim().isEmpty()) {
            throw new AppException("Title is required", HttpStatus.BAD_REQUEST);
        }
        if (request.getContent() == null || request.getContent().trim().isEmpty()) {
            throw new AppException("Content is required", HttpStatus.BAD_REQUEST);
        }

        Notice notice = Notice.builder()
                .title(request.getTitle().trim())
                .content(request.getContent().trim())
                .imageUrl(request.getImageUrl())
                .postedBy(user)
                .build();

        return toResponse(noticeRepository.save(notice));
    }

    @Transactional(readOnly = true)
    public List<NoticeResponse> getAllNotices() {
        // ✅ Only returns non-deleted notices
        return noticeRepository.findByDeletedFalseOrderByCreatedAtDesc()
                .stream().map(this::toResponse).collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public NoticeResponse getNoticeById(Long id) {
        // ✅ 404 if deleted or not found
        Notice notice = noticeRepository.findByIdAndDeletedFalse(id)
                .orElseThrow(() -> new ResourceNotFoundException("Notice not found: " + id));
        return toResponse(notice);
    }

    @Transactional
    public void deleteNotice(Long id, String email) {
        User user = getUser(email);
        if (user.getRole() != Role.ADMIN &&
            user.getRole() != Role.WARDEN &&
            user.getRole() != Role.CARETAKER) {
            throw new UnauthorizedException("Only admin, warden or caretaker can delete notices");
        }

        Notice notice = noticeRepository.findByIdAndDeletedFalse(id)
                .orElseThrow(() -> new ResourceNotFoundException("Notice not found: " + id));

        // ✅ Soft delete — mark as deleted, never remove from DB
        notice.setDeleted(true);
        notice.setDeletedAt(LocalDateTime.now());
        noticeRepository.save(notice);
    }

    private User getUser(String email) {
        return userRepository.findByEmail(email)
                .orElseThrow(() -> new ResourceNotFoundException("User not found: " + email));
    }

    private NoticeResponse toResponse(Notice n) {
        return NoticeResponse.builder()
                .id(n.getId())
                .title(n.getTitle())
                .content(n.getContent())
                .imageUrl(n.getImageUrl())
                .postedByName(n.getPostedBy().getName())
                .postedByRole(n.getPostedBy().getRole().name())
                .createdAt(n.getCreatedAt())
                .build();
    }
}

File: backend/src/services/notice.service.js
const prisma = require('../config/prisma');
const { NotFoundError, ForbiddenError, BadRequestError } = require('../utils/AppError');

const POSTER_ROLES = ['ADMIN', 'WARDEN', 'CARETAKER'];

function toResponse(n) {
  return {
    id: n.id,
    title: n.title,
    content: n.content,
    imageUrl: n.imageUrl,
    postedByName: n.poster.name,
    postedByRole: n.poster.role,
    createdAt: n.createdAt,
  };
}

async function getUserByEmail(email) {
  const user = await prisma.user.findUnique({ where: { email } });
  if (!user) throw new NotFoundError(`User not found: ${email}`);
  return user;
}

async function createNotice(email, request) {
  const user = await getUserByEmail(email);
  if (!POSTER_ROLES.includes(user.role)) {
    throw new ForbiddenError('Only admin, warden or caretaker can post notices');
  }
  if (!request.title || !request.title.trim()) throw new BadRequestError('Title is required');
  if (!request.content || !request.content.trim()) throw new BadRequestError('Content is required');

  const notice = await prisma.notice.create({
    data: {
      title: request.title.trim(),
      content: request.content.trim(),
      imageUrl: request.imageUrl,
      postedBy: user.id,
      deleted: false,
    },
    include: { poster: true },
  });

  return toResponse(notice);
}

async function getAllNotices() {
  const notices = await prisma.notice.findMany({
    where: { deleted: false },
    orderBy: { createdAt: 'desc' },
    include: { poster: true },
  });
  return notices.map(toResponse);
}

async function getNoticeById(id) {
  const notice = await prisma.notice.findFirst({
    where: { id: BigInt(id), deleted: false },
    include: { poster: true },
  });
  if (!notice) throw new NotFoundError(`Notice not found: ${id}`);
  return toResponse(notice);
}

async function deleteNotice(id, email) {
  const user = await getUserByEmail(email);
  if (!POSTER_ROLES.includes(user.role)) {
    throw new ForbiddenError('Only admin, warden or caretaker can delete notices');
  }

  const notice = await prisma.notice.findFirst({ where: { id: BigInt(id), deleted: false } });
  if (!notice) throw new NotFoundError(`Notice not found: ${id}`);

  await prisma.notice.update({
    where: { id: notice.id },
    data: { deleted: true, deletedAt: new Date() },
  });
}

module.exports = { createNotice, getAllNotices, getNoticeById, deleteNotice };
